test/test_qtile_cmd.py

```diff
@@ -0,0 +1,96 @@
+# Copyright (c) 2020 Guangwang Huang
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import pytest
+
+import libqtile.confreader
+import libqtile.config
+import libqtile.layout
+import libqtile.bar
+import libqtile.widget
+from libqtile.ipc import Client
+from libqtile.command_client import InteractiveCommandClient
+from libqtile.command_interface import IPCCommandInterface
+from libqtile.scripts.qtile_cmd import get_object, run_function
+
+
+class ServerConfig:
+    auto_fullscreen = True
+    keys = []
+    mouse = []
+    groups = [
+        libqtile.config.Group("a"),
+        libqtile.config.Group("b"),
+        libqtile.config.Group("c"),
+    ]
+    layouts = [
+        libqtile.layout.Stack(num_stacks=1),
+        libqtile.layout.Stack(num_stacks=2),
+        libqtile.layout.Stack(num_stacks=3),
+    ]
+    floating_layout = libqtile.layout.floating.Floating()
+    screens = [
+        libqtile.config.Screen(
+            bottom=libqtile.bar.Bar(
+                [
+                    libqtile.widget.TextBox(name="one"),
+                ],
+                20
+            ),
+        ),
+        libqtile.config.Screen(
+            bottom=libqtile.bar.Bar(
+                [
+                    libqtile.widget.TextBox(name="two"),
+                ],
+                20
+            ),
+        )
+    ]
+    main = None
+
+
+server_config = pytest.mark.parametrize("qtile", [ServerConfig], indirect=True)
+
+
+@server_config
+def test_qtile_cmd(qtile):
+
+    def _qtile_cmd(obj_spec, function, args):
+        "Imitate the behavior of qtile_cmd"
```

Comment:
Is it impossible to use `qtile_cmd` directly?

Reply:
You mean by calling the script directly? It's possible but I'm afraid it's a bit slower, and the output need to be parsed again if some fields need to be asserted.

Reply:
Yes but then we test the whole thing and I feel it has a lot of value, more than the speed gain.

Reply:
Yeah, I agree that the correctness of the WM and also the test is more important, so I changed it to call `qtile-cmd` directly now.